### PR TITLE
Add statistics to output

### PR DESCRIPTION
**Pull request by swashcap**
We’ll eventually merge the _addstatisticoutput_ branch to enable better result reporting for the computation. This PR gets the conversation started.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -56,6 +56,7 @@
     "jstat": "^1.5.2",
     "lodash": "^4.11.1",
     "numeric": "^1.2.6",
-    "pify": "^2.3.0"
+    "pify": "^2.3.0",
+    "distributions":"^1.0.0"
```
Comment (swashcap) [inline package.json]: Going to re-add the _distributions_ package via the CLI. This will modify _package.json_.

File: src/index.js [hunk 2]
```diff
@@ -45,13 +48,213 @@ module.exports = {
...
       type: 'function',
-      fn(opts) { return localRunner.run(opts); },
+      fn(opts) { 
+      // if W doesn't merge, return run(), if W merge, return statistic()
+//        console.log('opts.remoteResult.data is',opts.remoteResult.data);
+//         console.log('opts.previousData is', opts.previousData);
+       console.log('opts.remoteResult.data.endOfIteration is' , opts.remoteResult.data.endOfIteration);
+       console.log('opts.remoteResult.data.statisticStep is', opts.remoteResult.data.statisticStep);
+        
+        if ((opts.remoteResult.data.endOfIteration == true) && (opts.remoteResult.data.statisticStep == 0)){
```
Comment (swashcap) [inline src/index.js]: I’m a little iffy on these branches based on remote result state. Can we separate these into different local computations and let COINSTAC’s pipeline runner do the work?

File: src/index.js [hunk 3]
```diff
@@ -45,13 +48,213 @@ module.exports = {
...
+          
+          // step 0 calculate local statistics and localYMean
+           const biasedX = opts.previousData.biasedX;
+           const y = opts.previousData.y;
+           const localCount = y.length;
+           const betaVector = regression.oneShot(biasedX, y);
+           const rSquared = regression.rSquared(biasedX, y, betaVector);
+           const tValue = regression.tValue(biasedX, y, betaVector);
+           var tdist = distributions.Studentt(localCount-1);
```
Comment (swashcap) [inline src/index.js]: `var`? Prefer `const` or `let`.

File: src/index.js [hunk 4]
```diff
@@ -45,13 +48,213 @@ module.exports = {
...
+           return {
+             betaVector,
+             localCount,
+             localMeanY,
+             rSquared,
+             tValue,
+             pValue,
+             biasedX,
+             y,
+             endOfIteration : true,
```
Comment (swashcap) [inline src/index.js]: Our style guide prefers no space before the `:` char:

```diff
-             endOfIteration : true,
+             endOfIteration: true,
```

File: src/index.js [hunk 5]
```diff
@@ -45,13 +48,213 @@ module.exports = {
...
+             statisticStep : 0,
+          };          
+        }
+        else if ((opts.remoteResult.data.endOfIteration == true) && (opts.remoteResult.data.statisticStep == 1)){
```
Comment (swashcap) [inline src/index.js]: The fact that we're branching on `statisticStep` makes me want to separate these into local computations, as mentioned before. We can at least move functionality into _src/local.js_ to clean up our computation file.

File: src/index.js [hunk 6]
```diff
@@ -45,13 +48,213 @@ module.exports = {
...
+
+         // calculate averageBetaVector
+         const currW = opts.previousData.currW;
+
+         // calculate globalMeanY
+         const siteCount=userResults.length;
+         var totalY=0;
+         var globalYCount=0;
+
+         for (let i=0; i < siteCount; i += 1) {
```
Comment (swashcap) [inline src/index.js]: An `Array#reduce` could clean this up a bit.

File: src/index.js [hunk 7]
```diff
@@ -45,13 +48,213 @@ module.exports = {
...
+
+     }
+    else if ((opts.userResults[0].data.endOfIteration == true) && (opts.userResults[0].data.statisticStep == 1)) {
+         // if statisticStep==1, remote node calculates the global statistics 
+         
+         const userResults = opts.userResults;
+
+        // get passed parameters from local nodes
+         const currW = opts.previousData.currW;
+         const betaVectorLocal = userResults.map(r => r.data.betaVector);
```
Comment (swashcap) [inline src/index.js]: There’s a lot of repeated iteration over `userResults`, [similar to _decentralized-single-shot-ridge-regression_](https://github.com/MRN-Code/decentralized-single-shot-ridge-regression/pull/11#discussion_r96543959).

File: src/index.js (hunk 2 above)
Comment (JingMing623) [inline src/index.js]: Yes, we could merge this part into the local.js, or we can have local_comp1.js, local_comp2.js....... and on the remote node, we can also have remote_comp1.js, local_comp2.js..... ,  maybe you can help me on this. I did plan to write separate code for each condition.......Probably show me an example of modularizing code, I think single shot also can be modularized

File: src/index.js (hunk 5 above)
Comment (JingMing623) [inline src/index.js]: Yes, we should separate the computation.

**Comment by swashcap** [comment]
I rebased this branch so it has upstream changes. You might need to do a hard reset when merging this branch with your local copy.

**Comment by JingMing623** [comment]
Okay. Will work on resetting tomorrow. Thanks Cory.

**Comment by swashcap** [comment]
b51b964217d10bfff47f95e19fc3f35ee2ce6692 changes the computation step separation into separate functions, unified into `remote.run` and `local.run` functions. Test coverage is low, but let's ensure it actually works first.

**Comment by swashcap** [comment]
One thing to note: I swapped all `console.log` calls with [debug](https://www.npmjs.com/package/debug) to make runs less chatty. To enable logging you’ll need to ensure the `DEBUG` env var is set:

```shell
$ export DEBUG=coinstac:multishot
$ npm run simulate
```

**Comment by swashcap** [comment]
This PR is pretty outdated. Closing, we can open a new one for refactoring if needed.

**Pull request closed without merging**